src/nvim/api/command.c

```diff
@@ -779,41 +770,41 @@ static void build_cmdline_str(char **cmdlinep, exarg_T *eap, CmdParseInfo *cmdin
 
   // Keep the index of the position where command name starts, so eap->cmd can point to it.
   size_t cmdname_idx = cmdline.size;
-  kv_printf(cmdline, "%s", eap->cmd);
+  kv_concat(cmdline, eap->cmd);
 
   // Command bang.
   if (eap->argt & EX_BANG && eap->forceit) {
-    kv_printf(cmdline, "!");
+    kv_concat(cmdline, "!");
   }
 
   // Command register.
   if (eap->argt & EX_REGSTR && eap->regname) {
     kv_printf(cmdline, " %c", eap->regname);
   }
 
-  // Iterate through each argument and store the starting index and length of each argument
-  size_t *argidx = xcalloc(argc, sizeof(size_t));
+  // As arguments are appended, use the command index and argument indices to set the
+  // values of eap->cmd, eap->arg and eap->args.
   eap->argc = argc;
   eap->arglens = xcalloc(argc, sizeof(size_t));
-  for (size_t i = 0; i < argc; i++) {
-    argidx[i] = cmdline.size + 1;  // add 1 to account for the space.
-    eap->arglens[i] = STRLEN(args[i]);
-    kv_printf(cmdline, " %s", args[i]);
-  }
-
-  // Now that all the arguments are appended, use the command index and argument indices to set the
-  // values of eap->cmd, eap->arg and eap->args.
-  eap->cmd = cmdline.items + cmdname_idx;
   eap->args = xcalloc(argc, sizeof(char *));
+  eap->cmd = cmdline.items + cmdname_idx;
+
   for (size_t i = 0; i < argc; i++) {
-    eap->args[i] = cmdline.items + argidx[i];
+    String s = args.items[i].data.string;
+    kv_concat(cmdline, " ");
+    eap->args[i] = cmdline.items + cmdline.size;
```

Comment:
`cmdline` can be reallocated in kv_concat & friends, so the saved pointer could be invalid.

Reply:
Yes, just realised that, thanks.

Reply:
Yep, don't worry. Already running this through the santizer myself.

Reply:
Don't revert that previous stuff, it seemed to be all good and won't fix this issue anyway. The problem is that cmdline can be reallocated inside the loop where you save pointers to cmdline.

My suggestion is to save the offset (cmdline size) in the cmdline just before iterating over the arguments, then append them to cmdline and after all concatenation is done and cmdline won't be resized anymore, iterate over them again and recreate pointers to arguments from the initial offset + string size of the argument.

Reply:
I did try exactly that, but what I had seemed a bit obfuscated and didn't really impact the benchmark.

Can you share a diff of what you think might work?

Reply:
```diff
diff --git a/src/nvim/api/command.c b/src/nvim/api/command.c
index 9b2447e72..8bff7c685 100644
--- a/src/nvim/api/command.c
+++ b/src/nvim/api/command.c
@@ -787,19 +787,26 @@ static void build_cmdline_str(char **cmdlinep, exarg_T *eap, CmdParseInfo *cmdin
   eap->argc = argc;
   eap->arglens = xcalloc(argc, sizeof(size_t));
   eap->args = xcalloc(argc, sizeof(char *));
-  eap->cmd = cmdline.items + cmdname_idx;
 
   for (size_t i = 0; i < argc; i++) {
     String s = args.items[i].data.string;
     kv_concat(cmdline, " ");
-    eap->args[i] = cmdline.items + cmdline.size;
-    eap->arglens[i] = s.size;
     kv_concat_len(cmdline, s.data, s.size);
   }
 
   // Done appending to cmdline, ensure it is NUL terminated
   kv_push(cmdline, NUL);
 
+  eap->cmd = cmdline.items + cmdname_idx;
+  size_t offset = cmdline.size;
+  for (size_t i = 0; i < argc; i++) {
+    String s = args.items[i].data.string;
+    offset++; // kv_concat(cmdline, " ");
+    eap->args[i] = cmdline.items + offset;
+    eap->arglens[i] = s.size;
+    offset += s.size;
+  }
+
   // If there isn't an argument, make eap->arg point to end of cmdline.
   eap->arg = argc > 0 ? eap->args[0] : cmdline.items + cmdline.size;
 
```

Reply:
Ah no, `size_t offset = cmdline.size;` is in the wrong place, it should be before the first loop.

Reply:
This is basically what I had but found it makes the code harder to read. I guess we can go with it.